lib/ecto/repo/schema.ex

```diff
@@ -76,6 +86,41 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
+  defp extract_header_and_fields(query = %Ecto.Query{}, _schema, _dumper, _autogen_id, _placeholder_map, adapter) do
+    case query.select do
+      %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}} ->
+        header = Map.new(args, fn {k, _} -> {k, true} end)
+
+        query = Map.update!(query, :select, fn select ->
+          Map.update!(select, :expr, fn {op, ctx, args} ->
+            args = Enum.sort_by(args, &elem(&1, 0))
+            {op, ctx, args}
+          end)
+        end)
```

Comment:
Quick question: is this sorting going to work if the Map has more than 32 elements? Then Map.keys/1 and this function may return different orders... perhaps it is best if this function return this `headers` as a list so we can just go with the order in `{:%{}, ..., args}`?

Reply:
Are maps not sorted when they have over 32 elements? That's good to know. I agree, this is kind of sketchy. I would also much rather use a list that has the same order as the select expression to generate the fields list in the INSERT INTO.

Reply:
Yeah, so let's return headers a list here from this function and remove the Map.keys from the caller.